pkg/queryfrontend/queryinstant_codec.go

```diff
@@ -283,22 +302,55 @@ func vectorMerge(resps []*queryrange.PrometheusInstantQueryResponse) *queryrange
 	if len(output) == 0 {
 		return &queryrange.Vector{
 			Samples: make([]*queryrange.Sample, 0),
-		}
+		}, nil
 	}
 
-	keys := make([]string, 0, len(output))
-	for key := range output {
-		keys = append(keys, key)
+	var ss []*queryrange.Sample
+	for _, v := range output {
+		ss = append(ss, v)
 	}
-	sort.Strings(keys)
 
+	sort.Slice(ss, func(i, j int) bool {
+		// Order is determined by the sortFn in the query.
+		if sortAsc {
+			return ss[i].Sample.Value < ss[j].Sample.Value
+		} else if sortDesc {
+			return ss[i].Sample.Value > ss[j].Sample.Value
+		} else {
+			// Fallback on sorting by series
+			m1 := cortexpb.FromLabelAdaptersToLabels(ss[i].Labels).String()
+			m2 := cortexpb.FromLabelAdaptersToLabels(ss[j].Labels).String()
+			return m1 < m2
+		}
+	})
 	result := &queryrange.Vector{
 		Samples: make([]*queryrange.Sample, 0, len(output)),
 	}
-	for _, key := range keys {
-		result.Samples = append(result.Samples, output[key])
-	}
-	return result
+	result.Samples = append(result.Samples, ss...)
+	return result, nil
+}
+
+func parseQueryForSort(q string) (bool, bool, error) {
+	expr, err := promqlparser.ParseExpr(q)
+	var sortAsc bool = false
+	var sortDesc bool = false
+	promqlparser.Inspect(expr, func(n promqlparser.Node, _ []promqlparser.Node) error {
+		if call, ok := n.(*promqlparser.Call); ok {
+			sortFn := promqlparser.Functions["sort"]
+			sortDescFn := promqlparser.Functions["sort_desc"]
+			if call.Func.Name == sortFn.Name {
```

Comment:
We can just do `call.Fun.Name == "sort"` or `call.Fun.Name == "sort_desc"`.
Better to make them constants as well.